packages/pages-shared/asset-server/handler.ts

```diff
@@ -70,12 +70,20 @@ type ServeAsset<AssetEntry> = (
 	options?: { preserve: boolean }
 ) => Promise<Response>;
 
+type CacheStatus = "hit" | "miss";
+type CacheResult<A extends string> = `${A}-${CacheStatus}`;
+export type HandlerMetrics = {
+	preservationCacheResult?: CacheResult<"checked">;
+	earlyHintsResult?: CacheResult<"used" | "notused">;
+};
+
 type FullHandlerContext<AssetEntry, ContentNegotiation, Asset> = {
 	request: Request;
 	metadata: Metadata;
 	xServerEnvHeader?: string;
 	xDeploymentIdHeader?: boolean;
 	logError: (err: Error) => void;
+	setMetrics: (metrics: HandlerMetrics) => void;
```

Comment:
We should probably make this optional, so we can just avoid passing it in with `wrangler pages dev`'s usage.

Reply:
Ah good call, will update